In [3]:
import pandas as pd
import env
import numpy as np

1. Run python -m pip install pymysql from your terminal to install the mysql client (any folder is fine)
2. cd into your exercises folder for this module and run echo env.py >> .gitignore
3. Create a function named get_db_url.
 It should accept a username, hostname, password, and database name and return a url connection string formatted like in the example at the start of this lesson.
4. Use your function to obtain a connection to the employees database

In [4]:
url = env.get_db_url('employees')


5. Once you have successfully run a query:

a. Intentionally make a typo in the database url. What kind of error message do you see?



b. Intentionally make an error in your SQL query. What does the error message look like?

6. Read the employees and titles tables into two separate DataFrames.

In [15]:
employees=pd.read_sql('SELECT * FROM employees', url)

titles=pd.read_sql('SELECT * FROM titles', url)

employees, titles



(        emp_no  birth_date first_name last_name gender   hire_date
 0        10001  1953-09-02     Georgi   Facello      M  1986-06-26
 1        10002  1964-06-02    Bezalel    Simmel      F  1985-11-21
 2        10003  1959-12-03      Parto   Bamford      M  1986-08-28
 3        10004  1954-05-01  Chirstian   Koblick      M  1986-12-01
 4        10005  1955-01-21    Kyoichi  Maliniak      M  1989-09-12
 ...        ...         ...        ...       ...    ...         ...
 300019  499995  1958-09-24     Dekang  Lichtner      F  1993-01-12
 300020  499996  1953-03-07       Zito      Baaz      M  1990-09-27
 300021  499997  1961-08-03    Berhard    Lenart      M  1986-04-21
 300022  499998  1956-09-05   Patricia   Breugel      M  1993-10-13
 300023  499999  1958-05-01     Sachin   Tsukuda      M  1997-11-30
 
 [300024 rows x 6 columns],
         emp_no            title   from_date     to_date
 0        10001  Senior Engineer  1986-06-26  9999-01-01
 1        10002            Staff  1996-0

7. How many rows and columns do you have in each DataFrame? Is that what you expected?

 employees
 [300024 rows x 6 columns]

 titles
 [443308 rows x 4 columns])



8. Display the summary statistics for each DataFrame.

In [18]:
employees.describe() , titles.describe()

(              emp_no
 count  300024.000000
 mean   253321.763392
 std    161828.235540
 min     10001.000000
 25%     85006.750000
 50%    249987.500000
 75%    424993.250000
 max    499999.000000,
               emp_no
 count  443308.000000
 mean   253075.034430
 std    161853.292613
 min     10001.000000
 25%     84855.750000
 50%    249847.500000
 75%    424891.250000
 max    499999.000000)

9. How many unique titles are in the titles DataFrame?

In [23]:
unique_title_query= '''SELECT title
                    FROM titles 
                    GROUP BY title;'''


unique_titles = pd.read_sql(unique_title_query, url)
len(unique_titles)

7

10. What is the oldest date in the to_date column?

In [35]:
oldest_to_date_query ='''SELECT to_date
                        FROM titles
                        ORDER BY to_date 
                        LIMIT 1;'''



oldest_to_date= pd.read_sql(oldest_to_date_query,url)

oldest_to_date

to_date
0  1985-03-01

11.  What is the most recent date in the to_date column?

In [36]:
most_recent_query ='''SELECT to_date
                        FROM titles
                        ORDER BY to_date DESC
                        LIMIT 1;'''



recent_to_date= pd.read_sql(most_recent_query,url)

recent_to_date

to_date
0  9999-01-01

EXERCISES II

1. Copy the users and roles DataFrames from the examples above.